base/sysinfo.jl

```diff
@@ -459,6 +459,9 @@ for executable permissions only (with `.exe` and `.com` extensions added on
 Windows platforms); no searching of `PATH` is performed.
 """
 function which(program_name::String)
+    if isempty(program_name)
```

Comment:
Let's nip things like `" "` in the bud as well; do `isempty(strip(program_name))`

Reply:
So it turns out that this already works with the current code (at least on Linux).  

The reason is that ```isexecutable("/usr/bin/ ") == false``` while ```isexecutable("/usr/bin/") == true```.

It makes sense to add the test though.

Reply:
In that case, I think it makes more sense to change `isexecutable()` to `isfile() && isexecutable()`.  What do you think?

Reply:
Yes, I agree that makes sense.

I think we might as well keep the early exit for an empty string though.  There's no point in searching a bunch of directories for something they can't possibly contain.

Reply:
While "" isn't a valid file name, " " is perfectly legal (if a bit insane) as an executable name.

Reply:
Yep, though I think it's OK to include a test for that because the odds of there being such a file on a test system is very low.

There are currently no other tests for Sys.which and in general it's a bit complicated to write them because it would require controlling environment variables as well as the filesystem.  However testing that Sys.which(" ") == nothing is easy and almost certain to be right.

Reply:
FWIW, it's not so hard to make bulletproof tests for this:

```
mktempdir() do temp_dir
	my_bin = joinpath(temp_dir, "my_bin")
	touch(my_bin)
	chmod(my_bin, 0o755)
	withenv("PATH" => temp_dir) do
		@test abspath(Sys.which("my_bin")) == abspath(my_bin)
		@test Sys.which(" ") == nothing
		@test Sys.which("") == nothing
	end
end
```

Something like that should work, modulo `chmod()` issues on Windows.  (E.g. don't try to do test creating non-executable files on windows.... everything is executable on windows from Julia's perspective ;) )

Reply:
We already have tests for that (in test/spawn.jl), we just need tests for the new code (empty string case).

Reply:
Great.  Good to merge from my perspective then.

Reply:
@vtjnash 

I'm not clear on why the tests are there.  I thought the convention was that functions in base/foo.jl were tested in test/foo.jl.  

But since that's where the other which tests are I should probably move the new ones into test/spawn.jl which will also allow using the existing testing framework to ensure that the blank filename test is correct.